### PR TITLE
windowTime: fixes and tests

### PR DESCRIPTION
**Pull request by staltz**
Fix windowTime() operator to dispose window Subjects when the destination Subscriber is
unsubscribed.

Also allow a Subscription to add a child Subscription parameterized by `<any>`, not by `<T>`. There isn't a clear reason why all child Subscriptions should be type `<T>`, and there are cases where we need to
have a `<any>` child.

### CODE REVIEW

**Comment by blesh** [comment]
Yup. This looks good. Thanks for catching this with the `window` operators.

**Comment by staltz** [comment]
Cool. Next PR coming up also for `windowWhen`.

File: spec/operators/windowTime-spec.js [hunk 1]
```diff
@@ -123,6 +131,37 @@ describe('Observable.prototype.windowTime', function () {
...
+    virtualTimeScheduler.schedule(function () { source.next('e'); }, 50);
+    virtualTimeScheduler.schedule(function () { source.next('f'); }, 60);
+    virtualTimeScheduler.schedule(function () { source.next('g'); }, 70);
+    virtualTimeScheduler.schedule(function () {
+      try {
+        window.subscribe();
+      }
+      catch (err) {
+        expect(err.message).toBe('Cannot subscribe to a disposed Subject.');
+        done();
```
Comment (kwonoj) [inline spec/operators/windowTime-spec.js]: should this test be async test while using virtualtimescheduler? locally tried sync assertion as other PR (https://github.com/ReactiveX/RxJS/pull/1012/files#diff-f0b7f1f59929d2619cbf53504f594616R219) seems works.

Comment (blesh) [inline spec/operators/windowTime-spec.js]: VirtualTimeScheduler is the basis for TestScheduler. You should be able to just `flush()` the whole thing synchronously. ... I'm more curious about why this is in here.

File: spec/operators/windowTime-spec.js [hunk 2]
```diff
@@ -123,6 +131,37 @@ describe('Observable.prototype.windowTime', function () {
...
+      }
+    }, 80);
+
+    virtualTimeScheduler.flush();
```
Comment (blesh) [inline spec/operators/windowTime-spec.js]: Yeah, if you make assertions after this point, it doesn't need to be an async test.

File: spec/operators/windowTime-spec.js [hunk 3]
```diff
@@ -123,6 +131,37 @@ describe('Observable.prototype.windowTime', function () {
     expectObservable(result, unsub).toBe(expected, values);
   });
 
+  it('should dispose window Subjects if the outer is unsubscribed early', function (done) {
+    var virtualTimeScheduler = new Rx.VirtualTimeScheduler();
```
Comment (blesh) [inline spec/operators/windowTime-spec.js]: in all synchronous tests, there's a global `rxTestScheduler` you can use to do this work. It's based off of `VirtualTimeScheduler`, and you don't have to call `flush` on it per-say. (but you could).

Comment (blesh) [inline spec/operators/windowTime-spec.js]: @staltz is there any reason this can't be done with standard marble tests?

Comment (staltz) [inline spec/operators/windowTime-spec.js]: So, my standard approach was initially marble tests of course. Then I got into problems where I wanted to subscribe to a window (an inner Observable of the outer result) **AFTER** the whole outer is unsubscribed manually. I didn't figure out how to do that with marble diagrams without creating additional weirdness around it. I can give it another try. (the same applies to all other windowVariation operator tests).

**Comment by staltz** [comment]
@blesh I updated this PR to use `rxTestScheduler` with marble tests, not `virtualTimeScheduler`. I can use the same approach with other variants of window.

**Comment by blesh** [comment]
@staltz ... perfect. That's exactly the right thing to do there. You can schedule anything you like on `rxTestScheduler` and it will `flush` before making the assertions.

File: spec/operators/windowTime-spec.js [hunk 4]
```diff
@@ -108,19 +116,45 @@ describe('Observable.prototype.windowTime', function () {
...
+    expectObservable(result, unsub).toBe(expected, values);
+    expectSubscriptions(source.subscriptions).toBe(sourceSubs);
+    rxTestScheduler.schedule(function () {
+      try {
+        window.subscribe();
+      }
+      catch (err) {
+        expect(err.message).toBe('Cannot subscribe to a disposed Subject.');
+      }
+    }, 150);
```
Comment (kwonoj) [inline spec/operators/windowTime-spec.js]: Will this test also to use `time` in https://github.com/ReactiveX/RxJS/pull/1032 in later?

Comment (staltz) [inline spec/operators/windowTime-spec.js]: Yes.

Comment (kwonoj) [inline spec/operators/windowTime-spec.js]: Thanks for clarification. Then let'a make #1032 land first, update this PR.

**Comment by blesh** [comment]
Merged with b73e260. Thanks @staltz

**Pull request closed without merging**